Assets/MixedRealityToolkit/Utilities/MixedRealityServiceRegistry.cs

```diff
@@ -0,0 +1,222 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit
+{
+    /// <summary>
+    /// Static class that represents the Mixed Reality Toolkit service registry.
+    /// </summary>
+    /// <remarks>
+    /// The service registry is used to enable discovery of and access to active Mixed Reality Toolkit services at
+    /// runtime without requiring direct code reference to a singleton style component.
+    /// </remarks>
+    public static class MixedRealityServiceRegistry
+    {
+        // todo: is this the data type we want?
+        /// <summary>
+        /// The service registry store where the key is the Type of the service interface and the value is
+        /// a pair in which they key is the service instance and the value is the regisrar instance.
+        /// </summary>
+        private static Dictionary<Type, List<KeyValuePair<IMixedRealityService, IMixedRealityServiceRegistrar>>> registry = 
+            new Dictionary<Type, List<KeyValuePair<IMixedRealityService, IMixedRealityServiceRegistrar>>>();
+
+        /// <summary>
+        /// Static constructor.
+        /// </summary>
+        static MixedRealityServiceRegistry()
+        { }
+
+        /// <summary>
+        /// Adds an <see cref="IMixedRealityService"/> instance to the registry.
+        /// </summary>
+        /// <typeparam name="T">The interface type of the service being added.</typeparam>
+        /// <param name="serviceInstance">Instance of the service to add.</param>
+        /// <param name="registrar">Instance of the registrar manages the service.</param>
+        /// <returns>
+        /// True if the service was successfully added, false otherwise.
+        /// </returns>
+        public static bool AddService<T>(T serviceInstance, IMixedRealityServiceRegistrar registrar) where T : IMixedRealityService
+        {
+            Type interfaceType = typeof(T);
+            T existingService;
+
+            if (TryGetService<T>(out existingService, serviceInstance.Name))
+            {
+                return false;
+            }
```

Comment:
Do we want to depend on callers to filter out dataproviders from coming here? i.e. I see MixedRealityToolkit doing things where it filters out data providers, but if this is something we want to prevent from ever happening it would probably be good to enforce at a lower level.

Reply:
Good call!